spring-cloud-gcp-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/integration/PubSubHeaderMapper.java

```diff
@@ -0,0 +1,83 @@
+/*
+ *  Copyright 2018 original author or authors.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.pubsub.integration;
+
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+
+import org.springframework.messaging.MessageHeaders;
+import org.springframework.messaging.support.HeaderMapper;
+import org.springframework.util.Assert;
+
+public class PubSubHeaderMapper implements HeaderMapper<Map<String, String>> {
+
+	/**
+	 * Headers to exclude in the {@link #fromHeaders(MessageHeaders, Map)}.
+	 */
+	private String[] doNotMapHeaderNames = {
+			MessageHeaders.ID, MessageHeaders.TIMESTAMP
+	};
+
+	private boolean filterHeaders;
+
+	/**
+	 * Set the names of the headers to filtered out in {@link #fromHeaders(MessageHeaders, Map)}.
+	 * @param doNotMapHeaderNames
+	 */
+	public void setDoNotMapHeaderNames(String... doNotMapHeaderNames) {
+		Assert.notNull(doNotMapHeaderNames, "Header names can't be null.");
+		Assert.noNullElements(doNotMapHeaderNames, "No header name can be null.");
+		this.doNotMapHeaderNames = doNotMapHeaderNames;
```

Comment:
`Arrays.copyOf(doNotMapHeaderNames, doNotMapHeaderNames.length);` must be used to avoid the cases when end-user may mutate the original array.
On the other hand SonarQube doesn't like this code - exactly for the reason I've explained.